proposals/0000-custom-executors.md

```diff
@@ -0,0 +1,464 @@
+# Custom Executors
+
+* Proposal: [SE-NNNN](NNNN-custom-executors.md)
+* Authors: [John McCall](https://github.com/rjmccall)
+* Review Manager: TBD
+* Status: **Awaiting implementation**
+
+## Introduction
+
+It is sometimes important to control exactly how code is
+executed. This proposal lays out a system of custom executors
+which can schedule and run opaque jobs, and it describes
+how actors and tasks can be directed to run on a specific
+executor.
+
+## Motivation
+
+Swift's concurrency design is intentionally vague about the
+details of how code is actually run.  Most code does not rely
+on specific properties of the execution environment, such as
+being run to a specific operating system thread, and instead
+needs only on high-level semantic properties, such as that
+no other code will be accessing certain variables concurrently.
+Maintaining flexibility about how work is scheduled onto threads
+allows Swift to avoid certain performance pitfalls by default.
+
+Nonetheless, it is sometimes useful to more finely control how
+code is executed:
+
+- The code may need to cooperate with an existing system that
+  expects to run code in a certain way.
+
+  For example, the system might expect certain kinds of work
+  to be scheduled in special ways, like how some platforms
+  require UI code to be run on the main thread, or how Apple's
+  Core Data framework requires code that touches managed objects
+  to be performed by the managed object context.
+
+  For another example, a project might have a large amount
+  of existing code which protects some state with a shared queue.
+  In principle, this is the actor pattern, and the code could be
+  rewritten to use Swift's actor support. However, it may be
+  impossible to do that, or at least impractical to do it
+  immediately. Using the existing queue as the executor for an
+  actor allows code to adopt actors more incrementally.
+
+- The code may depend on being run on a specific system thread.
+
+  For example, some libraries maintain state in thread-local
+  variables, and running code on the wrong thread will lead to
+  broken assumptions in the library.
+
+  For another example, not all execution environments are
+  homogenous; some threads may be pinned to processors with
+  extra capabilities.
+
+- The code's performance may benefit from the programmer being
+  more explicit about where code should run.
+
+  For example, if one actor frequently makes requests of another,
+  and the actors rarely benefit from running concurrently,
+  configuring them to use the same executor may decrease the
+  runtime costs of switching between them.
+
+  For another example, if an asynchronous function makes many
+  calls to the same actor without any intervening suspensions,
+  running the function explicitly on that actor's executor may
+  allow Swift to avoid a lot of switching overhead (or may even
+  be necessary to perform those calls "atomically").
```

Comment:
Thanks @rjmccall ! If you want another example: Just as you describe below, many systems do have fixed-width thread pools that don't block as long as there is outstanding work to be done. If there's no outstanding work to be done, these threads do usually block (as you describe below). So without custom executors, implementing one of these systems (typically used for I/O) would also not be possible without prohibitive thread switching (because we're obviously can't block say a thread of the default executor in `kqueue`, `epoll_wait`, iouring, or `mach_msg`.

Or in other words: Any system that implements its custom I/O and also wants to run `async` code will need to use custom executors for that.

Reply:
To be clear, and I do remember that the proposal discusses this: while you certainly *could* use custom executors for that, I'm not sure it's a good idea, because having two fixed-width thread pools in a process basically means having one poorly-managed thread pool with way too many threads.  Abstractly, I think we really want to encourage people who are writing scales-to-the-number-of-cores thread pools to either use the default concurrent executor or replace it globally.  But if you think having multiple pools is important, I'd love to understand how you hope to use the second pool, and particularly how you think it should affect thread allocation in the main pool.

Reply:
The default mode in SwiftNIO is exactly what you describe: One scales-to-the-number-of-cores thread pool that is shared by everything. However, we leave this up to the user because not every application is homogenous. 

For example, let's say your server is supposed to -- at peak times -- `accept` as many connections as possible. How you'd achieve that is to use one thread pool with number-of-cores threads which each bind on the same socket address (with `SO_RESUEPORT`). You could also give those threads elevated priority in the OS and possibly even pin them to an individual hardware CPU core. This will work well but is still quite latency sensitive because the OS's "listen queue" (queue of incoming connections that has yet to be `accept`ed) is bounded. So if at the peak time you can't call `accept` fast enough, the OS will drop connections for you (once the listen queue is exhausted). So far, this works really well with just one replaced default pool. But imagine, in the same program, we also run other code which is also supposed to run but isn't supposed to interfere with the accept-as-fast-as-possible work. Imagine for example, some JSON is supposed to be decoded or some something like that. It could be a pretty big if the JSON decoding would now cause us to return to the scheduler less often. This may cause us to drop connections. Priorities can't solve this issue because they are not preemptive.

As a really simple fix, users can configure a second (fixed-width) thread pool on which they run the other work. They are basically ring fencing the two workloads that are running in the same address space from each other by having separate pools.

Other examples include
- if you were to run two separate systems that perform I/O in one process. Let's say on Linux one system that multiplexes I/O using `io_uring` and another one that multiplexes using `epoll`. They can't share the same thread pool because their whole eventing and I/O mechanism is totally different. They could for example use two thread pools of size `number_of_cores/2` each. 
- in unit testing we usually start up and tear down the thread pool per test (suite) for them to be completely independent, work better when tests are run in parallel, and also to be able to test interesting scheduling scenarios with say 1 thread, or 3 threads, or some other "interesting" number
- say you have a service and an "admin interface". You may choose to have the admin interface run on a separate thread pool to be sure to never disrupt the main service from achieving its goals even if the admin interface does maybe do processing that's a little too heavy (say parse a JSON, protobuf, compress/decompress, ...) and may harm the latency of the main service